[16.0][IMP] membership_variable_period: enable membership renewal

flaenen opened the pull request:
A member's membership is now renewed when a new variable period membership product of the same category is purchased. The new period will begin the day after the end of the last period.

polchampion commented:
functional tests

**Backend and logic**
- [x] The "buy membership" button takes current and future membership lines into account when renewing variable memberships
- [x] If a past membership line has expired, renewal starts from today
- [x] Postponed renewal only applies to memberships from the same category
- [x] Postponed renewal only applies to membership states that grant effective membership ('paid', 'free' and 'invoiced')

**POS UI** [waiting for deployment before testing]
- [x] Postponed renewal applies when a membership is sold in the POS to an effective member
- [x] Compatibility with pos_membership_extension: test of both modules together

polchampion reviewed (approved):
[successful functional test](https://github.com/OCA/vertical-association/pull/205#issuecomment-2761450298)

remytms reviewed (approved):
LGTM.

polchampion commented:
@OCA/vertical-association-maintainers Can you take a look at this PR?

polchampion commented:
@pedrobaeza Hi Pedro, could you take a look at this? Our PR is awaiting approval :-)

pedrobaeza commented:
/ocabot migration membership_variable_period

Hi, a PSC should review this PR to get it merged. https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#review

I'm not able to review right now, sorry.

rafaelbn commented:
This is not a migration but a improvement / enhancement

What do you think about this @fcvalgar @Andrii9090 @chienandalu ?

chienandalu commented on membership_variable_period/models/account_move_line.py:
```diff
@@ -12,20 +12,29 @@
     _inherit = "account.move.line"
 
-    def _prepare_membership_line(self, move, product, price_unit, line_id, qty=1.0):
+    def _prepare_membership_line(self, move, product, price_unit, line, qty=1.0):
```
I'm not against changing this arg type, but doing so in stable code could break someone else extensions...

It also makes the commit meaningful content more difficult to follow.

flaenen replied:
I will modify the code to keep the arg type.

flaenen replied:
Done

chienandalu reviewed (commented):
Thanks @flaenen 

I'm not sure I quite understand the context or use case for this changes. Can you elaborate it a little bit?

flaenen commented:
> Thanks @flaenen
> 
> I'm not sure I quite understand the context or use case for this changes. Can you elaborate it a little bit?

In our use case, a customer can become a member for one year by purchasing a membership product (usually via POS). Membership begins on the date of purchase and can be renewed before expiration by purchasing the product again. Renewal must begin the day after the end of the current period (not the day of purchase) to avoid overlapping periods, ensuring that the customer receives a full additional year of membership.

flaenen committed:
[IMP] membership_variable_period: enable membership renewal
A member's membership is renewed when a new variable period membership product
of the same category is purchased. The new period will begin the day after the
end of the last period.

polchampion commented:
> > Thanks @flaenen
> > I'm not sure I quite understand the context or use case for this changes. Can you elaborate it a little bit?
> 
> In our use case, a customer can become a member for one year by purchasing a membership product (usually via POS). Membership begins on the date of purchase and can be renewed before expiration by purchasing the product again. Renewal must begin the day after the end of the current period (not the day of purchase) to avoid overlapping periods, ensuring that the customer receives a full additional year of membership.

@OCA/vertical-association-maintainers do think this can now be merged? thanks a lot for taking a final look :-)

jguenat reviewed (approved):
Thank you for the improvement. Very useful to renew membership with variable period before the end of the membership validity.
Code looks good

pedrobaeza reviewed (approved):
/ocabot merge minor
